### ISSUE TITLE
Reproducing PANDA Results

### ISSUE THREAD

**Issue by Ajaz-Ahmad** [arneschmidt/attention_gp#12]
Hello,

Thanks for sharing your work. I was trying to reproduce the result for PANDA on ISUP grade but the results are not so good. Can you share the slide ids you used for training and testing your method.

Thanks and Regards
Ajaz Ahmad

**Comment by arneschmidt** [comment]
Hi Ajaz!

Thanks for reaching out!
I will check the feature tables and if they are not too large I can upload them and provide you access. This would be the best way to reproduce results.
Does that suit you?

Unfortunately, I am really busy right now so I hope I can work on that at the beginning of next week.

Best
Arne
________________________________
Von: Ajaz Ahmad ***@***.***>
Gesendet: Freitag, 24. Februar 2023 10:41
An: arneschmidt/attention_gp ***@***.***>
Cc: Subscribed ***@***.***>
Betreff: [arneschmidt/attention_gp] Reproducing PANDA Results (Issue #12)


Hello,

Thanks for sharing your work. I was trying to reproduce the result for PANDA on ISUP grade but the results are not so good. Can you share the slide ids you used for training and testing your method.

Thanks and Regards
Ajaz Ahmad

**Comment by Ajaz-Ahmad** [comment]
Hi Arne,

Thanks for responding. I understand those are large datasets. Anything is fine for me, but it will be good if I can reproduce feature creation step as well. So, if you can only share the bag names and labels that would do for now. 

I have tried creating features using resnet18 pretrained on imagenet over 600 Radboud quality checked slides. Ran with default configuration and then evaluated on Karolinska's 60 slides (quality checked) and here is the result I get
![image](https://user-images.githubusercontent.com/29792736/221150516-e13daeab-4f70-4085-a674-47dc5162f313.png)
Any idea what is not right here apart from different staining.
Thanks for your help

Kind Regards

**Comment by arneschmidt** [comment]
Ahh I see!
Actually, I just looked it up and found the corresponding table.
I will attach it to this mail.

Best
Arne
________________________________
Von: Ajaz Ahmad ***@***.***>
Gesendet: Freitag, 24. Februar 2023 11:00
An: arneschmidt/attention_gp ***@***.***>
Cc: arneschmidt ***@***.***>; Comment ***@***.***>
Betreff: Re: [arneschmidt/attention_gp] Reproducing PANDA Results (Issue #12)


Hi Arne,

Thanks for responding. I understand those are large datasets. Anything is fine for me, but it will be good if I can reproduce feature creation step as well. So, if you can only share the bag names and labels that would do for now.
Thanks for your help

Kind Regards

**Comment by Ajaz-Ahmad** [comment]
Also any idea on the result I shared

**Comment by arneschmidt** [comment]
So just with the results, it is hard to say what is the issue. The network used in the article two extract features has two major differences to your feature extraction network:

- it was pre-trained on histopathological images
- it is larger (EfficientNetB5)

In general, you can try the normal (deterministic) attention mechanism first, if this one doesn't work either, chances are high that Attention GP will not work in the given setting.

**Comment by Ajaz-Ahmad** [comment]
Hi Again

Based on your comment, what histopathological images were used to train the feature extractor? How did you train the feature extractor for PANDA?

**Comment by arneschmidt** [comment]
Hello, 
Sorry for the late reply! We used the feature extractor of a previous article (https://ieeexplore.ieee.org/document/9681818).
It was trained in the MIL setting on the Panda dataset. It uses the global labels to assign pseudo-labels for instances with high probabilities.
Best
Arne

**Comment by Ajaz-Ahmad** [comment]
Hi Again,

I have looked into the above paper and repository. I generated the patch csv files for prostate dataset. I tried config settings to mil and ssl models. But looking into the model architecture, what I understand is there is a standard feature extractor based on imagenet followed by dense layer with number of classes 4. The model focuses on patch which has confidence interval of 0.95% to be selected for strong augmentation.
But my question is regarding the feature extractor training which is utilized in Probabilistic Attention based on Gaussian Processes for Deep Multiple Instance Learning.

Can you please help?

Thanks in advance

**Comment by arneschmidt** [comment]
Hello,

I will share the configuration file to produce the features with MIL labels. I hope this helps to reproduce the feature extraction process.
Anyway, the exact reproduction of Panda patching, preprocessing, feature extractor training, testing, and predicting is quite difficult because it includes so many complex steps. Therefore, the exact experiment can better be reproduced with the already extracted features, I will also share them.

The above mentioned files can be found here:
https://drive.google.com/drive/folders/1S7flTOehdrTlx-YkBOvboEjc0M-wrzLs?usp=sharing

Hope I could help.
Best,
Arne

**Comment by Ajaz-Ahmad** [comment]
Thanks Arne,

Thanks fir sharing the test features, Can you also provide training features as well or are you uploading them still as it would be quite large. Otherwise, if you can share model's weights I can use it to extract features myself.

Thanks for your help

**Comment by arneschmidt** [comment]
I am glad I could help!
The features are still uploading, train and val will be available soon.
I can additionally upload the model in the next days.
Best,
Arne

**Comment by Ajaz-Ahmad** [comment]
Thanks Arne.
Can you also confirm my understanding here. You are trying to gain benefit from marked patches and apply weak augmentation on it. In the next step you select the patches which are unlabelled from Karolinska's positively labeled slide and predict on it. In the next step, you select the highly scored patches from unmarked patches and then if gt for the patch is available you use it and where not present you use pseudo labels predicted in the previous step. In the final stage to train MIL model, you supply wsi labels along with pseudo labels to train the model.
Am I correct here or did I miss any step?
Thanks

**Comment by arneschmidt** [comment]
The feature extractor is trained with the following strategy:
From the global labels, the occurrences of certain patch classes can be deducted. For example, a slide with primary Gleason grade 3 and secondary Gleason grade 4 contains patches of grades 3 and 4. The model assigns pseudo-labels of grades 3 and 4 for the 5 patches with the highest predicted probabilities in the corresponding classes. For healthy slides we know that all patches are healthy.
In the original paper, we also use some cancerous patch labels. But as we present a MIL model here, we train the model without any patch labels in this case. Just by the pseudo labels, the model learns step by step to classify the patches. This model does not achieve a super good classification performance in the MIL setting, but the feature extraction apparently works quite well.

So, we use these features for the probabilistic attention model to obtain the reported high results.
